qir/qat/Rules/Factory.cpp.orig

```diff
@@ -0,0 +1,801 @@
+
+namespace microsoft
+{
+namespace quantum
+{
+    namespace
+    {
+        using Instruction  = llvm::Instruction;
+        using Value        = llvm::Value;
+        using Builder      = ReplacementRule::Builder;
+        using Captures     = ReplacementRule::Captures;
+        using Replacements = ReplacementRule::Replacements;
+
+    } // namespace
+
+    using ReplacementRulePtr = RuleFactory::ReplacementRulePtr;
+    using namespace microsoft::quantum::notation;
+
+    RuleFactory::RuleFactory(
+        RuleSet&             rule_set,
+        AllocationManagerPtr qubit_alloc_manager,
+        AllocationManagerPtr result_alloc_manager,
+        ILoggerPtr           logger)
+      : rule_set_{rule_set}
+      , qubit_alloc_manager_{std::move(qubit_alloc_manager)}
+      , result_alloc_manager_{std::move(result_alloc_manager)}
+      , logger_{std::move(logger)}
+    {
+    }
+
+    void RuleFactory::usingConfiguration(FactoryConfiguration const& config)
+    {
+        default_integer_width_ = config.defaultIntegerWidth();
+
+        if (config.disableReferenceCounting())
+        {
+            disableReferenceCounting();
+        }
+
+        if (config.disableAliasCounting())
+        {
+            disableAliasCounting();
+        }
+
+        if (config.disableStringSupport())
+        {
+            disableStringSupport();
+        }
+
+        if (config.disableRecordOutputSupport())
+        {
+            disableRecordOutputSupport();
+        }
+
+        if (config.optimizeQuantumConstants())
+        {
+            optimizeConstantResult();
+        }
+
+        if (config.optimizeResultOne())
+        {
+            optimizeResultOne();
+        }
+
+        if (config.optimizeResultZero())
+        {
+            optimizeResultZero();
+        }
+
+        if (config.useStaticQubitArrayAllocation())
+        {
+            useStaticQubitArrayAllocation();
+        }
+
+        if (config.useStaticQubitAllocation())
+        {
+            useStaticQubitAllocation();
+        }
+
+        if (config.useStaticResultAllocation())
+        {
+            useStaticResultAllocation();
+        }
+
+        if (config.removeGetZeroOrOne())
+        {
+            removeGetZeroOrOne();
+        }
+    }
+
+    void RuleFactory::removeFunctionCall(String const& name)
+    {
+        addRule({callByNameOnly(name), deleteInstruction()});
+    }
+
+    void RuleFactory::removeGetZeroOrOne()
+    {
+        addRule(
+            {callByNameOnly("__quantum__rt__result_get_zero"), deleteUnusedInstruction()},
+            RuleSet::ReplaceDirection::ReplaceBackwards);
+        addRule(
+            {callByNameOnly("__quantum__rt__result_get_one"), deleteUnusedInstruction()},
+            RuleSet::ReplaceDirection::ReplaceBackwards);
+    }
+
+    void RuleFactory::resolveConstantArraySizes()
+    {
+        /// Array access replacement
+        auto size_replacer = [](Builder&, Value* val, Captures& cap, Replacements& replacements)
```

Comment:
Did you really intend the L115 be a _doc_ comment? (`///` as opposed to `//`)
Multiple occurrences.